---
title: "Bioconductor Quiz 3"
author: "dillonchewwx"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
---

# Question 1

**What is the mean expression across all features for sample 5 in the ALL dataset (from the ALL package)?**

```{r q1-setup, message=FALSE, warning=FALSE}
## Run this the first time to install the "ALL" package.
# BiocManager::install("ALL")
library(ALL)
data(ALL)
exprs(ALL[1:10,1:10])
```
We just need to get the mean expression for sample 5 and this can be done easily using the mean function. 

```{r q1-ans}
mean(exprs(ALL[,5]))
```

# Question 2
**We will use the biomaRt package to annotate an Affymetrix microarray. We want our results in the hg19 build of the human genome and we therefore need to connect to Ensembl 75 which is the latest release on this genome version. How to connect to older versions of Ensembl is described in the biomaRt package vignette; it can be achieved with the command `mart <- useMart(host='feb2014.archive.ensembl.org', biomart = "ENSEMBL_MART_ENSEMBL")`. Using this version of Ensembl, annotate each feature of the ALL dataset with the Ensembl gene id. How many probesets (features) are annotated with more than one Ensembl gene id?**

```{r q2, warning=FALSE, message=FALSE}
## Run this the first time to install the "biomaRt" package.
# BiocManager::install("biomaRt")
library(biomaRt)
mart<-useMart(host='feb2014.archive.ensembl.org', biomart = "ENSEMBL_MART_ENSEMBL")
ensembl<-useDataset("hsapiens_gene_ensembl", mart)

## Annotate the data
annotation(ALL)
featurenames<-featureNames(ALL)
annotation_ALL<-getBM(attributes=c("ensembl_gene_id", "affy_hg_u95av2"),
      filters="affy_hg_u95av2",
      values=featurenames,
      mart=ensembl)

library(tidyverse)
annotation_ALL %>%
    group_by(affy_hg_u95av2) %>%
    summarize(Count=n()) %>%
    filter(Count>1)
```

There are a total of 1045 features which are annotated with > 1 Ensembl gene id. 

# Question 3
**How many probesets (Affymetrix IDs) are annotated with one or more genes on the autosomes (chromosomes 1 to 22)?**

```{r q3}
attributes<-listAttributes(ensembl)
filters<-listFilters(ensembl)

# Get probesets on autosomes
annotation_ALL_autosomes<-getBM(attributes=c("ensembl_gene_id", "affy_hg_u95av2", "chromosome_name"),
      filters=c("affy_hg_u95av2", "chromosome_name"),
      values=list(featurenames, chromosome=1:22),
      mart=ensembl)

head(annotation_ALL_autosomes)
```

From here, we just calculate the number of unique probesets since the question asks for annotation with one or more genes.

```{r q3-answer}
annotation_ALL_autosomes$affy_hg_u95av2 %>% 
    unique() %>% 
    length()
```

# Question 4
**Use the MsetEx dataset from the minfiData package. Part of this question is to use the help system to figure out how to address the question. What is the mean value of the Methylation channel across the features for sample “5723646052_R04C01”?**

```{r q4-setup, eval=FALSE}
## Run this the first time to install the "minfiData" package.
BiocManager::install("minfiData")
library(minfiData)
```

Had some issues with installing this package and so made a guess. Answer is 7228.277.

# Question 5
**Access the processed data from NCBI GEO Accession number GSE788. What is the mean expression level of sample GSM9024?**

```{r q5, message=FALSE}
## Run this the first time to install the "GEOquery" package.
# BiocManager::install("GEOquery")
## Run this line before loading GEOquery - https://github.com/seandavi/GEOquery/issues/114
readr::local_edition(1)

library(GEOquery)
eList<-getGEO("GSE788")
eList_expr<-exprs(eList[[1]])
head(eList_expr)

mean(eList_expr[,2])
```

# Question 6
**We are using the airway dataset from the airway package. What is the average of the average length across the samples in the expriment?**

```{r q6, message=FALSE}
## Run this the first time to install the "airway" package.
# BiocManager::install("airway")
library(airway)
data(airway)
airway_df<-as.data.frame(colData(airway))
mean(airway_df$avgLength)
```

# Question 7
**We are using the airway dataset from the airway package. The features in this dataset are Ensembl genes. What is the number of Ensembl genes which have a count of 1 read or more in sample SRR1039512?**

```{r q7}
head(assay(airway))
sum(assay(airway)[,3]>=1)
```

# Question 8
**The airway dataset contains more than 64k features. How many of these features overlaps with transcripts on the autosomes (chromosomes 1-22) as represented by the TxDb.Hsapiens.UCSC.hg19.knownGene package? Clarification: A feature has to overlap the actual transcript, not the intron of a transcript.  So you will need to make sure that the transcript representation does not contain introns.**

```{r q8, message=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb<-TxDb.Hsapiens.UCSC.hg19.knownGene
txdb_exons<-exons(txdb)

# Select only autosomes
txdb_exons_autosomes<-keepStandardChromosomes(txdb_exons, pruning.mode="coarse") %>%
  dropSeqlevels(c("chrX", "chrY", "chrM"), pruning.mode="coarse")

txdb_exons_ncbi<-renameSeqlevels(txdb_exons_autosomes, mapSeqlevels(seqlevels(txdb_exons), style="NCBI"))
overlap<-subsetByOverlaps(airway, txdb_exons_ncbi)
dim(overlap)
```

# Question 9
**The expression measures of the airway dataset are the number of reads mapping to each feature. In the previous question we have established that many of these features do not overlap autosomal transcripts from the TxDb.Hsapiens.UCSC.hg19.knownGene. But how many reads map to features which overlaps these transcripts? Question: For sample SRR1039508, how big a percentage (expressed as a number between 0 and 1) of the total reads in the airway dataset for that sample, are part of a feature which overlaps an autosomal TxDb.Hsapiens.UCSC.hg19.knownGene transcript?**

```{r q9}
SRR1039508<-airway[,1]
SRR1039508_autosome<-subsetByOverlaps(SRR1039508, txdb_exons_ncbi)

SRR1039508_autosome_reads<-sum(assay(SRR1039508_autosome, "counts"))
total_reads<-sum(assay(SRR1039508, "counts"))

SRR1039508_autosome_reads/total_reads
```

# Question 10
Consider sample SRR1039508 and only consider features which overlap autosomal transcripts from TxDb.Hsapiens.UCSC.hg19.knownGene. We should be able to very roughly divide these transcripts into expressed and non expressed transcript. Expressed transcripts should be marked by H3K4me3 at their promoter. The airway dataset have assayed “airway smooth muscle cells”. In the Roadmap Epigenomics data set, the E096 is supposed to be “lung”. Obtain the H3K4me3 narrowPeaks from the E096 sample using the AnnotationHub package.

Question: What is the median number of counts per feature (for sample SRR1039508) containing a H3K4me narrowPeak in their promoter (only features which overlap autosomal transcripts from TxDb.Hsapiens.UCSC.hg19.knownGene are considered)?

Clarification: We are using the standard 2.2kb default Bioconductor promoter setting.

Conclusion Compare this to the median number of counts for features without a H3K4me3 peak. Note that this short analysis has not taken transcript lengths into account and it compares different genomic regions to each other; this is highly susceptible to bias such as sequence bias.

```{r q10, message=FALSE, warning=FALSE}
library(AnnotationHub)
# Get AH30596
ah<-AnnotationHub()
ah_E096<-query(ah, c("E096", "narrowPeak", "H3K4me3"))
ah_E096_record<-ah[["AH30596"]]

# Keep autosomes
ah_E096_record_autosome<-keepStandardChromosomes(ah_E096_record, pruning.mode="coarse") %>%
  dropSeqlevels(c("chrX", "chrY", "chrM"), pruning.mode="coarse")
ah_E096_record_ncbi<-renameSeqlevels(ah_E096_record_autosome,
                                     mapSeqlevels(seqlevels(txdb_exons), style="NCBI"))

# Change to NCBI Format
ncbi_group<-extractSeqlevelsByGroup(species="Homo sapiens", style="NCBI", group="auto")
sample_ncbi<-keepSeqlevels(range(rowRanges(SRR1039508_autosome)), ncbi_group)

overlap<-subsetByOverlaps(promoters(sample_ncbi), ah_E096_record_ncbi)
overlap<-subsetByOverlaps(SRR1039508, overlap)

median(assay(overlap, "counts"))
```

This answer appears to be much smaller than the options provided. Might have to recheck the procedure. 