---
title: "Bioconductor Quiz 1"
author: "dillonchewwx"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
---
# Quiz 1 Introduction

In this assessment we will study features of the human genome, version "hg19". We will only consider data which has been mapped to the autosomes (chr 1 to 22).

We will use data from the H1 cell line as assayed and quantified by the Roadmap Epigenomics project. The Roadmap Epigenomics project code for the H1 cell line is E003. For histone modification data, the Roadmap project makes several types of quantification available. We will use the so-called narrowPeak quantification.

Note: This entire quiz is one long analysis, so later questions refer to results generated in earlier questions.

Some biology: Bivalent chromatin is marked by a combination of active and repressive histone marks. A number of slightly different definitions exists; we will say a region is bivalent if it is enriched in both H3K4me3 and H3K27me3. Note that histone modification marks does not have a strand.

Bivalent chromatin has especially been considered in embryonic stem cells. An example of such a cell is the ENCODE Tier 1 cell line called H1. The Roadmap Epigenomics id for this cell line is "E003".

We will examine the relationship between bivalent chromatin and CpG Islands. CpG Islands are clusters of many CpG (this is juts CG dinucleotides). Several definitions exists of what is an “CpG Island”; we will use the UCSC definition. Because the CG dinucleotide is its own reverse complement, a CpG cluster exists on the forward strand if and only if it exists on the reverse strand. In other words, CpG Islands does not have a strand.

## Question 1

**Use the AnnotationHub package to obtain data on "CpG Islands" in the human genome. How many islands exists on the autosomes?**

```{r Load Packages, message=FALSE, warning=FALSE}
library(AnnotationHub)
library(GenomicRanges)
library(rtracklayer)
library(magrittr)
```

```{r q1-1}
ah<-AnnotationHub()
ah_human<-subset(ah, species=="Homo sapiens")
ah_human_cpg<-query(ah_human, c("CpG Islands", "hg19"))
ah_human_cpg
```

Here we can see that there is only 1 record of "AH5086".

```{r q1-2}
ah_human_cpg_ah5086<-ah_human_cpg[["AH5086"]]
unique(seqnames(ah_human_cpg_ah5086))
```

Note that there are additional chromosomes apart from the 23 found in human. We will have to subset the data to only keep those from chr1-chr22 as the question asks for autosomes. 

```{r q1-3, message=FALSE, warning=FALSE}
ah_human_cpg_ah5086_autosome<-ah_human_cpg_ah5086 %>%
    keepStandardChromosomes(pruning.mode="coarse") %>%
    dropSeqlevels(c("chrX", "chrY", "chrM"), pruning.mode="coarse")
seqlevels(ah_human_cpg_ah5086_autosome)
ah_human_cpg_ah5086_autosome
```

This returns 26641 ranges, which suggests that there are 26641 islands on the autosomes.

## Question 2

**How many CpG Islands exists on chromosome 4?**

```{r q2}
ah_human_cpg_ah5086_chr4<-ah_human_cpg_ah5086_autosome %>%
    keepSeqlevels(c("chr4"), pruning.mode="coarse")
ah_human_cpg_ah5086_chr4
```

Using the same method as before, we obtain 1031 CpG islands on chromosome 4. 

## Question 3

**Obtain the data for the H3K4me3 histone modification for the H1 cell line from Epigenomics Roadmap, using AnnotationHub. Subset these regions to only keep regions mapped to the autosomes (chromosomes 1 to 22). How many bases does these regions cover?**

```{r q3-1}
ah_human_H3K4me3<-query(ah_human, c("H3K4me3", "E003", "narrow"))
ah_human_H3K4me3
```

Using the descriptions in the introduction, we narrow down the search to one record. 

```{r q3-2}
ah_human_H3K4me3_AH29884<-ah_human_H3K4me3[["AH29884"]]
ah_human_H3K4me3_AH29884_autosome<-ah_human_H3K4me3_AH29884 %>%
    keepStandardChromosomes(pruning.mode="coarse") %>%
    dropSeqlevels(c("chrX", "chrY", "chrM"), pruning.mode="coarse")
sum(width(ah_human_H3K4me3_AH29884_autosome))
```

Using the same method to filter autosomes, the find that the number of bases is `r sum(width(ah_human_H3K4me3_AH29884_autosome))`.

## Question 4

**Obtain the data for the H3K27me3 histone modification for the H1 cell line from Epigenomics Roadmap, using the AnnotationHub package. Subset these regions to only keep regions mapped to the autosomes. In the return data, each region has an associated "signalValue". What is the mean signalValue across all regions on the standard chromosomes?**

```{r q4}
ah_human_H3K27me3<-query(ah_human, c("H3K27me3", "E003", "narrow"))
ah_human_H3K27me3
ah_human_H3K27me3_AH29892<-ah_human_H3K27me3[["AH29892"]]
ah_human_H3K27me3_AH29892_autosome<-ah_human_H3K27me3_AH29892 %>%
    keepStandardChromosomes(pruning.mode="coarse") %>%
    dropSeqlevels(c("chrX", "chrY", "chrM"), pruning.mode="coarse")
mean(ah_human_H3K27me3_AH29892_autosome$signalValue)
```

Same methods used as before to get the answer of `r mean(ah_human_H3K27me3_AH29892_autosome$signalValue)`.

## Question 5

**Bivalent regions are bound by both H3K4me3 and H3K27me3. Using the regions we have obtained above, how many bases on the standard chromosomes are bivalently marked?**

```{r q5}
bivalent_region<-intersect(ah_human_H3K4me3_AH29884_autosome, ah_human_H3K27me3_AH29892_autosome)
sum(width(bivalent_region))
```

## Question 6

**We will examine the extent to which bivalent regions overlap CpG Islands. How big a fraction (expressed as a number between 0 and 1) of the bivalent regions, overlap one or more CpG Islands?**

```{r q6}
bivalent_cpg_overlap<-findOverlaps(bivalent_region, ah_human_cpg_ah5086_autosome)
bivalent_cpg_overlap
length(unique(queryHits(bivalent_cpg_overlap)))/length(bivalent_region)
```

## Question 7

**How big a fraction (expressed as a number between 0 and 1) of the bases which are part of CpG Islands, are also bivalent marked?**

```{r q7}
cpg_bivalent_marked<-intersect(ah_human_cpg_ah5086_autosome, bivalent_region)
sum(width(cpg_bivalent_marked))/sum(width(ah_human_cpg_ah5086_autosome))
```

## Question 8

**How many bases are bivalently marked within 10kb of CpG Islands? Tip: consider using the "resize()" function.**

```{r q8}
bivalent_10kb<-resize(ah_human_cpg_ah5086_autosome, width=20000+width(ah_human_cpg_ah5086_autosome), fix="center")
cpg_bivalent_10kb<-intersect(bivalent_10kb, bivalent_region)
sum(width(cpg_bivalent_10kb))
```

## Question 9

**How big a fraction (expressed as a number between 0 and 1) of the human genome is contained in a CpG Island? Tip 1: the object returned by AnnotationHub contains "seqlengths". Tip 2: you may encounter an integer overflow. As described in the session on R Basic Types, you can address this by converting integers to numeric before summing them, "as.numeric()".**

```{r q9}
sum(width(ah_human_cpg_ah5086_autosome))/sum(seqlengths(ah_human_cpg_ah5086_autosome))
```

## Question 10

**Compute an odds-ratio for the overlap of bivalent marks with CpG islands.**

```{r q10}
overlapMat<-matrix(0, ncol=2, nrow=2)
rownames(overlapMat)<-c("bivalent_in", "bivalent_out")
colnames(overlapMat)<-c("CpG_in", "CpG_out")
overlapMat[1,1]<-sum(width(cpg_bivalent_marked))
overlapMat[1,2]<-sum(width(setdiff(bivalent_region, ah_human_cpg_ah5086_autosome)))
overlapMat[2,1]<-sum(width(setdiff(ah_human_cpg_ah5086_autosome, bivalent_region)))
overlapMat[2,2]<-sum(seqlengths(ah_human_cpg_ah5086_autosome))-sum(overlapMat)
overlapMat
oddsRatio<-overlapMat[1,1] * overlapMat[2,2] / (overlapMat[2,1] * overlapMat[1,2])
oddsRatio
```

Follow the method shown in the lecture. 

# Session Info

```{r sesinfo}
sessionInfo()
```