<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Kasper D. Hansen" />


<title>Usecase - Basic GRanges and AnnotationHub</title>


<style type="text/css">code{white-space: pre;}</style>
<link href="data:text/css;charset=utf-8,pre%20%2Eoperator%2C%0Apre%20%2Eparen%20%7B%0Acolor%3A%20rgb%28104%2C%20118%2C%20135%29%0A%7D%0Apre%20%2Eliteral%20%7B%0Acolor%3A%20%23990073%0A%7D%0Apre%20%2Enumber%20%7B%0Acolor%3A%20%23099%3B%0A%7D%0Apre%20%2Ecomment%20%7B%0Acolor%3A%20%23998%3B%0Afont%2Dstyle%3A%20italic%0A%7D%0Apre%20%2Ekeyword%20%7B%0Acolor%3A%20%23900%3B%0Afont%2Dweight%3A%20bold%0A%7D%0Apre%20%2Eidentifier%20%7B%0Acolor%3A%20rgb%280%2C%200%2C%200%29%3B%0A%7D%0Apre%20%2Estring%20%7B%0Acolor%3A%20%23d14%3B%0A%7D%0A" rel="stylesheet" type="text/css" />
<script src="data:application/x-javascript;base64,
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();

"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link href="data:text/css;charset=utf-8,body%2C%20td%20%7B%0Afont%2Dfamily%3A%20sans%2Dserif%3B%0Abackground%2Dcolor%3A%20white%3B%0Afont%2Dsize%3A%2013px%3B%0A%7D%0Abody%20%7B%0Amax%2Dwidth%3A%20800px%3B%0Amargin%3A%200%20auto%3B%0Apadding%3A%201em%201em%202em%3B%0Aline%2Dheight%3A%2020px%3B%0A%7D%0A%0Ap%2C%20pre%20%7B%20margin%3A%200em%200em%201em%3B%0A%7D%0A%0Aimg%2C%20table%20%7B%0Amargin%3A%200em%20auto%201em%3B%0A%7D%0Ap%20%7B%0Atext%2Dalign%3A%20justify%3B%0A%7D%0Att%2C%20code%2C%20pre%20%7B%0Afont%2Dfamily%3A%20%27DejaVu%20Sans%20Mono%27%2C%20%27Droid%20Sans%20Mono%27%2C%20%27Lucida%20Console%27%2C%20Consolas%2C%20Monaco%2C%20monospace%3B%0A%7D%0Ah1%2C%20h2%2C%20h3%2C%20h4%2C%20h5%2C%20h6%20%7B%20font%2Dfamily%3A%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Amargin%3A%201%2E2em%200em%200%2E6em%200em%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%20250%25%3B%0Afont%2Dweight%3A%20normal%3B%0Acolor%3A%20%2387b13f%3B%0Aline%2Dheight%3A%201%2E1em%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%20160%25%3B%0Afont%2Dweight%3A%20normal%3B%0Aline%2Dheight%3A%201%2E4em%3B%0Aborder%2Dbottom%3A%201px%20%231a81c2%20solid%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%20130%25%3B%20%7D%0Ah2%2C%20h3%20%7B%0Acolor%3A%20%231a81c2%3B%0A%7D%0Ah4%2C%20h5%2C%20h6%20%7B%0Afont%2Dsize%3A115%25%3B%0A%7D%20%0A%0Aa%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Aactive%20%7B%20outline%3A%20none%3B%20%7D%0Aa%3Avisited%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Ahover%20%7B%20color%3A%20%234c94c2%3B%20%7D%0Apre%2C%20img%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Adisplay%3A%20block%3B%0A%7D%0Apre%20%7B%0Aborder%3A%200px%20none%3B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0Awhite%2Dspace%3A%20pre%3B%0Aoverflow%2Dx%3A%20auto%3B%0A%7D%0Apre%20code%20%7B%0Aborder%3A%201px%20%23aaa%20dashed%3B%0Abackground%2Dcolor%3A%20white%3B%0Adisplay%3A%20block%3B%0Apadding%3A%201em%3B%20color%3A%20%23111%3B%0Aoverflow%2Dx%3A%20inherit%3B%0A%7D%0A%0Apre%20code%5Bclass%5D%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Apre%5Bclass%5D%20code%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Acode%20%7B%20background%2Dcolor%3A%20transparent%3B%0Acolor%3A%20%2387b13f%3B%0Afont%2Dsize%3A%2092%25%3B%0A%7D%0A%0Atable%2C%20td%2C%20th%20%7B%0Aborder%3A%20none%3B%0Apadding%3A%200%200%2E5em%3B%0A%7D%0A%0Atbody%20tr%3Anth%2Dchild%28odd%29%20td%20%7B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0A%7D%0Ablockquote%20%7B%0Acolor%3A%23666666%3B%0Amargin%3A0%3B%0Apadding%2Dleft%3A%201em%3B%0Aborder%2Dleft%3A%200%2E5em%20%23EEE%20solid%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200px%3B%0Aborder%2Dbottom%3A%20none%3B%0Aborder%2Dtop%2Dwidth%3A%20thin%3B%0Aborder%2Dtop%2Dstyle%3A%20dotted%3B%0Aborder%2Dtop%2Dcolor%3A%20%23999999%3B%0A%7D%0A%40media%20print%20%7B%0A%2A%20%7B%0Abackground%3A%20transparent%20%21important%3B%0Acolor%3A%20black%20%21important%3B%0Afilter%3Anone%20%21important%3B%0A%2Dms%2Dfilter%3A%20none%20%21important%3B%0A%7D%0Abody%20%7B%0Afont%2Dsize%3A12pt%3B%0Amax%2Dwidth%3A100%25%3B%0A%7D%0Aa%2C%20a%3Avisited%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Avisibility%3A%20hidden%3B%0Apage%2Dbreak%2Dbefore%3A%20always%3B%0A%7D%0Apre%2C%20blockquote%20%7B%0Apadding%2Dright%3A%201em%3B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Atr%2C%20img%20%7B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%20%21important%3B%0A%7D%0A%40page%20%3Aleft%20%7B%0Amargin%3A%2015mm%2020mm%2015mm%2010mm%3B%0A%7D%0A%40page%20%3Aright%20%7B%0Amargin%3A%2015mm%2010mm%2015mm%2020mm%3B%0A%7D%0Ap%2C%20h2%2C%20h3%20%7B%0Aorphans%3A%203%3B%20widows%3A%203%3B%0A%7D%0Ah2%2C%20h3%20%7B%0Apage%2Dbreak%2Dafter%3A%20avoid%3B%0A%7D%0A%7D%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Usecase - Basic GRanges and AnnotationHub</h1>
<h4 class="author"><em>Kasper D. Hansen</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#dependencies">Dependencies</a></li>
<li><a href="#corrections">Corrections</a></li>
<li><a href="#overview">Overview</a></li>
<li><a href="#getting-the-data">Getting the data</a></li>
<li><a href="#sessioninfo">SessionInfo</a></li>
</ul>
</div>

<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>This document has the following dependencies:</p>
<pre class="r"><code>library(GenomicRanges)
library(rtracklayer)
library(AnnotationHub)</code></pre>
<p>Use the following commands to install these packages in R.</p>
<pre class="r"><code>source(&quot;http://www.bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;GenomicRanges&quot;, &quot;rtracklayer&quot;, &quot;AnnotationHub&quot;))</code></pre>
</div>
<div id="corrections" class="section level2">
<h2>Corrections</h2>
<p>Improvements and corrections to this document can be submitted on its <a href="https://github.com/kasperdanielhansen/genbioconductor/blob/master/Rmd/Usecase_AnnotationHub_GRanges.Rmd">GitHub</a> in its <a href="https://github.com/kasperdanielhansen/genbioconductor">repository</a>.</p>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>We are going to use <em><a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a></em> and <em><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a></em> to access ENCODE data on the H3K4me3 histone modification in a specific cell line. This histone modification is believed to mark active promoters, and we are going to attempt to verify this statement. This involves</p>
<ol style="list-style-type: decimal">
<li>Getting the ENCODE histone data using <em><a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a></em>.</li>
<li>Getting promoters using <em><a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a></em>.</li>
<li>Comparing the histone data and promoters using <code>findOverlaps</code> in <em><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a></em>.</li>
</ol>
</div>
<div id="getting-the-data" class="section level2">
<h2>Getting the data</h2>
<p>First we use <em><a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a></em> to get data on homo sapiens.</p>
<pre class="r"><code>ah &lt;- AnnotationHub()
ah &lt;- subset(ah, species == &quot;Homo sapiens&quot;)</code></pre>
<p>Next we search for two keywords: <strong>H3K4me3</strong> and <strong>Gm12878</strong> which is the name of the cell line we are interested in.</p>
<pre class="r"><code>qhs &lt;- query(ah, &quot;H3K4me3&quot;)
qhs &lt;- query(qhs, &quot;Gm12878&quot;)</code></pre>
<p>(Note: I like to keep my full annotation hub around, so I can re-do my query with a different search term in case I end up with no hits. This is why I start assigning output to the <code>qhs</code> object and not <code>ah</code>).</p>
<p>Lets have a look</p>
<pre class="r"><code>qhs</code></pre>
<pre><code>## AnnotationHub with 11 records
## # snapshotDate(): 2015-08-17 
## # $dataprovider: BroadInstitute, UCSC
## # $species: Homo sapiens
## # $rdataclass: GRanges, BigWigFile
## # additional mcols(): taxonomyid, genome, description, tags,
## #   sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH23256&quot;]]' 
## 
##             title                                                      
##   AH23256 | wgEncodeBroadHistoneGm12878H3k4me3StdPk.broadPeak.gz       
##   AH27075 | wgEncodeUwHistoneGm12878H3k4me3StdHotspotsRep1.broadPeak.gz
##   AH27076 | wgEncodeUwHistoneGm12878H3k4me3StdHotspotsRep2.broadPeak.gz
##   AH27077 | wgEncodeUwHistoneGm12878H3k4me3StdPkRep1.narrowPeak.gz     
##   AH27078 | wgEncodeUwHistoneGm12878H3k4me3StdPkRep2.narrowPeak.gz     
##   ...       ...                                                        
##   AH30747 | E116-H3K4me3.narrowPeak.gz                                 
##   AH31690 | E116-H3K4me3.gappedPeak.gz                                 
##   AH32869 | E116-H3K4me3.fc.signal.bigwig                              
##   AH33901 | E116-H3K4me3.pval.signal.bigwig                            
##   AH40294 | E116-H3K4me3.imputed.pval.signal.bigwig</code></pre>
<p>Note how some of these hits don’t contain <strong>Gm12878</strong> in their title. This is a useful illustration of how <code>query</code> searches over multiple fields.</p>
<p>Lets have a closer look at this</p>
<pre class="r"><code>qhs$title</code></pre>
<pre><code>##  [1] &quot;wgEncodeBroadHistoneGm12878H3k4me3StdPk.broadPeak.gz&quot;       
##  [2] &quot;wgEncodeUwHistoneGm12878H3k4me3StdHotspotsRep1.broadPeak.gz&quot;
##  [3] &quot;wgEncodeUwHistoneGm12878H3k4me3StdHotspotsRep2.broadPeak.gz&quot;
##  [4] &quot;wgEncodeUwHistoneGm12878H3k4me3StdPkRep1.narrowPeak.gz&quot;     
##  [5] &quot;wgEncodeUwHistoneGm12878H3k4me3StdPkRep2.narrowPeak.gz&quot;     
##  [6] &quot;E116-H3K4me3.broadPeak.gz&quot;                                  
##  [7] &quot;E116-H3K4me3.narrowPeak.gz&quot;                                 
##  [8] &quot;E116-H3K4me3.gappedPeak.gz&quot;                                 
##  [9] &quot;E116-H3K4me3.fc.signal.bigwig&quot;                              
## [10] &quot;E116-H3K4me3.pval.signal.bigwig&quot;                            
## [11] &quot;E116-H3K4me3.imputed.pval.signal.bigwig&quot;</code></pre>
<pre class="r"><code>qhs$dataprovider</code></pre>
<pre><code>##  [1] &quot;UCSC&quot;           &quot;UCSC&quot;           &quot;UCSC&quot;           &quot;UCSC&quot;          
##  [5] &quot;UCSC&quot;           &quot;BroadInstitute&quot; &quot;BroadInstitute&quot; &quot;BroadInstitute&quot;
##  [9] &quot;BroadInstitute&quot; &quot;BroadInstitute&quot; &quot;BroadInstitute&quot;</code></pre>
<p>This result is a great illustration of the mess of public data. It turns our that <strong>E116</strong> is a Roadmap Epigenetics code for the <strong>Gm12878</strong> cell line. The first 5 hits are from ENCODE, hosted at UCSC and the last 6 hits are from Roadmap Epigenomics hosted at the Broad Institute. The Roadmap data is different representation (and peaks) from the same underlying data. For the ENCODE data, two different centers did the same experiment in the same cell line (<code>Broad</code>, hit 1) and (<code>Uw</code>, hit 2-5), where <code>Uw</code> exposed data on two replicates (whatever that means). These two experiments seems to be analyzed using different algorithms. It is even possible that the Roadmap data is from the same raw data but just analyzed using different algorithms.</p>
<p>Lets take a look at the <code>narrowPeak</code> data:</p>
<pre class="r"><code>gr1 &lt;- subset(qhs, title == &quot;wgEncodeUwHistoneGm12878H3k4me3StdPkRep1.narrowPeak.gz&quot;)[[1]]
gr1</code></pre>
<pre><code>## GRanges object with 74470 ranges and 6 metadata columns:
##           seqnames                 ranges strand   |        name     score
##              &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;character&gt; &lt;numeric&gt;
##       [1]     chr1       [713301, 713450]      *   |        &lt;NA&gt;         0
##       [2]     chr1       [713501, 713650]      *   |        &lt;NA&gt;         0
##       [3]     chr1       [713881, 714030]      *   |        &lt;NA&gt;         0
##       [4]     chr1       [714181, 714330]      *   |        &lt;NA&gt;         0
##       [5]     chr1       [714481, 714630]      *   |        &lt;NA&gt;         0
##       ...      ...                    ...    ... ...         ...       ...
##   [74466]     chrX [154492741, 154492890]      *   |        &lt;NA&gt;         0
##   [74467]     chrX [154493401, 154493550]      *   |        &lt;NA&gt;         0
##   [74468]     chrX [154560441, 154560590]      *   |        &lt;NA&gt;         0
##   [74469]     chrX [154562121, 154562270]      *   |        &lt;NA&gt;         0
##   [74470]     chrX [154842061, 154842210]      *   |        &lt;NA&gt;         0
##           signalValue    pValue    qValue      peak
##             &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##       [1]          91  112.7680        -1        -1
##       [2]          25   26.7181        -1        -1
##       [3]          81   77.4798        -1        -1
##       [4]          32  106.5650        -1        -1
##       [5]         122  153.8320        -1        -1
##       ...         ...       ...       ...       ...
##   [74466]          43  52.20930        -1        -1
##   [74467]         122 203.16800        -1        -1
##   [74468]           8   4.49236        -1        -1
##   [74469]           8   4.41978        -1        -1
##   [74470]         125 170.20100        -1        -1
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>gr2 &lt;- subset(qhs, title == &quot;E116-H3K4me3.narrowPeak.gz&quot;)[[1]]
gr2</code></pre>
<pre><code>## GRanges object with 76188 ranges and 6 metadata columns:
##           seqnames                 ranges strand   |        name     score
##              &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;character&gt; &lt;numeric&gt;
##       [1]     chr9 [123553464, 123557122]      *   |      Rank_1       623
##       [2]     chr3 [ 53196213,  53197995]      *   |      Rank_2       622
##       [3]    chr18 [  9137534,   9142676]      *   |      Rank_3       583
##       [4]    chr11 [ 75110593,  75111943]      *   |      Rank_4       548
##       [5]    chr13 [ 41343776,  41345943]      *   |      Rank_5       543
##       ...      ...                    ...    ... ...         ...       ...
##   [76184]     chr8 [ 98881411,  98881613]      *   |  Rank_76184        20
##   [76185]     chr9 [ 26812360,  26812533]      *   |  Rank_76185        20
##   [76186]     chrX [ 49019816,  49020016]      *   |  Rank_76186        20
##   [76187]     chrX [ 55932872,  55933045]      *   |  Rank_76187        20
##   [76188]     chr5 [118322235, 118322408]      *   |  Rank_76188        20
##           signalValue    pValue    qValue      peak
##             &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##       [1]    23.09216  62.34132  52.85911      1736
##       [2]    24.91976  62.22835  52.85911       439
##       [3]    22.48938  58.39180  50.67302       442
##       [4]    22.47056  54.84914  47.77176       752
##       [5]    20.82804  54.31837  47.29083       909
##       ...         ...       ...       ...       ...
##   [76184]     2.63471   2.02539   0.27458        94
##   [76185]     2.63471   2.02539   0.27458        32
##   [76186]     2.63471   2.02539   0.27458        82
##   [76187]     2.63471   2.02539   0.27458        41
##   [76188]     2.62206   2.00970   0.26790        52
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
<p>(Note: I use this code, where I use <code>title</code> to refer to the different resources, to make this script more robust over time).</p>
<p>This gives us two <code>GRanges</code> objects. Let us look at the distribution of peak widths:</p>
<pre class="r"><code>summary(width(gr1))</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   150.0   150.0   150.0   150.3   150.0   410.0</code></pre>
<pre class="r"><code>table(width(gr1))</code></pre>
<pre><code>## 
##   150   210   230   250   270   290   390   410 
## 74313     1    13    24    37    77     4     1</code></pre>
<pre class="r"><code>summary(width(gr2))</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##     174     243     387     662     770   12340</code></pre>
<p>Turns out that almost all peaks in <code>gr1</code> have a width of 150bp, whereas <code>gr2</code> is much more variable. This is likely a product of the data processing algorithm; it can be very hard to figure out the details of this.</p>
<p>At this time one can spend a lot of time thinking about which datasets is best. We will avoid this (important) discussion; since we referred to ENCODE data above (in the Overview section), we will stick with <code>gr1</code>.</p>
<p>Now we need to get some promoter coordinates. There are multiple ways to do this in Bioconductor, but here I will do a quick lookup for RefSeq in my annotation hub. RefSeq is a highly curated (aka conservative) collection of genes.</p>
<p>Lets get started</p>
<pre class="r"><code>qhs &lt;- query(ah, &quot;RefSeq&quot;)
qhs</code></pre>
<pre><code>## AnnotationHub with 8 records
## # snapshotDate(): 2015-08-17 
## # $dataprovider: UCSC
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description, tags,
## #   sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH5040&quot;]]' 
## 
##            title       
##   AH5040 | RefSeq Genes
##   AH5041 | Other RefSeq
##   AH5155 | RefSeq Genes
##   AH5156 | Other RefSeq
##   AH5306 | RefSeq Genes
##   AH5307 | Other RefSeq
##   AH5431 | RefSeq Genes
##   AH5432 | Other RefSeq</code></pre>
<p>So this gives us multiple datasets, all with very similar names. We probably need the thing called <code>RefSeq Genes</code> and not <code>Other RefSeq</code> but why are there multiple resources with the same name?</p>
<p>Turns out the answer makes sense:</p>
<pre class="r"><code>qhs$genome</code></pre>
<pre><code>## [1] &quot;hg19&quot; &quot;hg19&quot; &quot;hg18&quot; &quot;hg18&quot; &quot;hg17&quot; &quot;hg17&quot; &quot;hg16&quot; &quot;hg16&quot;</code></pre>
<p>This looks like the same resources, but in different genome builds. We have</p>
<pre class="r"><code>genome(gr1)</code></pre>
<pre><code>##  chr1 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19  chr2 
##    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA 
## chr20 chr21 chr22  chr3  chr4  chr5  chr6  chr7  chr8  chr9  chrX 
##    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA</code></pre>
<p>so we know which one to get:</p>
<pre class="r"><code>refseq &lt;- qhs[qhs$genome == &quot;hg19&quot; &amp; qhs$title == &quot;RefSeq Genes&quot;]
refseq</code></pre>
<pre><code>## AnnotationHub with 1 record
## # snapshotDate(): 2015-08-17 
## # names(): AH5040
## # $dataprovider: UCSC
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # $title: RefSeq Genes
## # $description: GRanges object from UCSC track 'RefSeq Genes'
## # $taxonomyid: 9606
## # $genome: hg19
## # $sourcetype: UCSC track
## # $sourceurl: rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg19/dat...
## # $sourcelastmodifieddate: NA
## # $sourcesize: NA
## # $tags: refGene, UCSC, track, Gene, Transcript, Annotation 
## # retrieve record with 'object[[&quot;AH5040&quot;]]'</code></pre>
<pre class="r"><code>refseq &lt;- refseq[[1]] ## Downloads</code></pre>
<p>Lets have a look</p>
<pre class="r"><code>refseq</code></pre>
<pre><code>## UCSC track 'refGene'
## UCSCData object with 50066 ranges and 5 metadata columns:
##                        seqnames               ranges strand   |
##                           &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;   |
##       [1]                  chr1 [66999825, 67210768]      +   |
##       [2]                  chr1 [ 8378145,  8404227]      +   |
##       [3]                  chr1 [48998527, 50489626]      -   |
##       [4]                  chr1 [16767167, 16786584]      +   |
##       [5]                  chr1 [16767167, 16786584]      +   |
##       ...                   ...                  ...    ... ...
##   [50062] chr19_gl000209_random     [ 57209,  68123]      +   |
##   [50063] chr19_gl000209_random     [ 46646,  68123]      +   |
##   [50064] chr19_gl000209_random     [ 98135, 112667]      +   |
##   [50065] chr19_gl000209_random     [ 70071,  84658]      +   |
##   [50066] chr19_gl000209_random     [131433, 145745]      +   |
##                   name     score     itemRgb                thick
##            &lt;character&gt; &lt;numeric&gt; &lt;character&gt;            &lt;IRanges&gt;
##       [1]    NM_032291         0        &lt;NA&gt; [67000042, 67208778]
##       [2] NM_001080397         0        &lt;NA&gt; [ 8378169,  8404073]
##       [3]    NM_032785         0        &lt;NA&gt; [48999845, 50489468]
##       [4] NM_001145277         0        &lt;NA&gt; [16767257, 16785491]
##       [5] NM_001145278         0        &lt;NA&gt; [16767257, 16785385]
##       ...          ...       ...         ...                  ...
##   [50062]    NM_002255         0        &lt;NA&gt;     [ 57249,  67717]
##   [50063] NM_001258383         0        &lt;NA&gt;     [ 57132,  67717]
##   [50064]    NM_012313         0        &lt;NA&gt;     [ 98146, 112480]
##   [50065] NM_001083539         0        &lt;NA&gt;     [ 70108,  83979]
##   [50066]    NM_012312         0        &lt;NA&gt;     [131468, 145120]
##                                                     blocks
##                                              &lt;IRangesList&gt;
##       [1] [    1,   227] [91706, 91769] [98929, 98953] ...
##       [2]       [   1,  102] [6222, 6642] [7214, 7306] ...
##       [3]       [   1, 1439] [2036, 2062] [6788, 6884] ...
##       [4]       [   1,  182] [2961, 3061] [7199, 7303] ...
##       [5]       [   1,  104] [2961, 3061] [7199, 7303] ...
##       ...                                              ...
##   [50062]       [   1,   80] [ 280,  315] [1182, 1466] ...
##   [50063] [    1,    86] [10414, 10643] [10843, 10878] ...
##   [50064]       [   1,   46] [1523, 1557] [4002, 4301] ...
##   [50065]       [   1,   71] [1071, 1106] [1851, 2135] ...
##   [50066]       [   1,   69] [ 862,  897] [3334, 3633] ...
##   -------
##   seqinfo: 93 sequences from hg19 genome</code></pre>
<p>Let us look at the number of isoforms per gene name:</p>
<pre class="r"><code>table(table(refseq$name))</code></pre>
<pre><code>## 
##     1     2     3     4     5     6     7     8     9    10    12    13 
## 45178   568   102    27    74    52   171   129    13    19     1     1 
##    19 
##     5</code></pre>
<p>(the <code>table(table())</code> construction may seems weird at first, but its a great way to get a quick tabular summary of occurrences with the same name). This shows that almost all genes have a single transcript, which reflects how conservative RefSeq is.</p>
<p>Notice that each transcript is a separate range. Therefore, each transcript is represented as the “outer” coordinates of the gene; the ranges does not exclude introns. Because we got this from UCSC I happen to know that the <code>$blocks</code> metadata really contains the coordinates of the different exons. In a later example we will introduce a gene representation where we keep track of exons, introns and transcripts from the same gene, through something known as a <code>TxDb</code> (transcript database) object.</p>
<p>For now, we just need the promoters, so we don’t really care about introns. We need to keep track of which strand each transcript is on, to get the transcription start site.</p>
<p>Or we could just use the convenience function <code>promoters()</code>:</p>
<pre class="r"><code>promoters &lt;- promoters(refseq)
table(width(promoters))</code></pre>
<pre><code>## 
##  2200 
## 50066</code></pre>
<pre class="r"><code>args(promoters)</code></pre>
<pre><code>## NULL</code></pre>
<p>There are many definitions of promoters based on transcription start site. The default in this function is to use 2kb upstream and 200bp downstream of the start site. Let’s keep this.</p>
<p>Now we compute which promoters have a H3K4me3 peak in them:</p>
<pre class="r"><code>ov &lt;- findOverlaps(promoters, gr1)
ov</code></pre>
<pre><code>## Hits object with 46022 hits and 0 metadata columns:
##           queryHits subjectHits
##           &lt;integer&gt;   &lt;integer&gt;
##       [1]         2         387
##       [2]         3        2523
##       [3]         4         774
##       [4]         5         774
##       [5]         6        1114
##       ...       ...         ...
##   [46018]     47457       47491
##   [46019]     47459       47493
##   [46020]     47459       47494
##   [46021]     47460       47493
##   [46022]     47460       47494
##   -------
##   queryLength: 50066
##   subjectLength: 74470</code></pre>
<p>Let us compute how many percent of the peaks are in a promoter</p>
<pre class="r"><code>length(unique(queryHits(ov))) / length(gr1)</code></pre>
<pre><code>## [1] 0.3333691</code></pre>
<p>and how many percent of promoters have a peak in them</p>
<pre class="r"><code>length(unique(subjectHits(ov))) / length(promoters)</code></pre>
<pre><code>## [1] 0.440878</code></pre>
<p>My rule of thumb is that any cell type has at most 50% of genes expressed, which fits well with these numbers. We also see that there are many H3K4me3 peaks which do not lie in a genic promoter. This is actually expected.</p>
<p>Now, the overlap is non-empty, but is it interesting or “significant”. Answering this question thoroughly requires thinking deeply about background distributions etc. We will avoid a careful statistical approach to this question, and will instead do a few back-of-the-envelope calculations to get a feel for it.</p>
<p>First we notice that both promoters and peaks are small compared to the size of the human genome:</p>
<pre class="r"><code>sum(width(reduce(gr1))) / 10^6</code></pre>
<pre><code>## [1] 11.19044</code></pre>
<pre class="r"><code>sum(width(reduce(promoters))) / 10^6</code></pre>
<pre><code>## [1] 64.3005</code></pre>
<p>(result is in megabases; the human genome is 3000 megabases). Just this fact alone should convince you that the overlap is highly unlikely happen purely by chance. Let us calculate the overlap in megabases:</p>
<pre class="r"><code>sum(width(intersect(gr1, promoters))) / 10^6</code></pre>
<pre><code>## [1] 0</code></pre>
<p>That’s weird; why is the overlap empty when we know (using <code>findOverlaps</code>) that there is plenty of overlap? Turns out the answer is strand. We need to ignore the strand when we do this calculation:</p>
<pre class="r"><code>sum(width(intersect(gr1, promoters, ignore.strand = TRUE))) / 10^6</code></pre>
<pre><code>## [1] 3.019608</code></pre>
<p>This brings us back to the question of the size of the promoter set above; this is also affected by strand when there is a promoter on each strand which overlaps. Contrast</p>
<pre class="r"><code>sum(width(reduce(promoters))) / 10^6</code></pre>
<pre><code>## [1] 64.3005</code></pre>
<pre class="r"><code>sum(width(reduce(promoters, ignore.strand = TRUE))) / 10^6</code></pre>
<pre><code>## [1] 62.27957</code></pre>
<p>Strand can be troublesome!</p>
<p>Let us compute a small 2x2 matrix for which bases are in promoters and/or peaks:</p>
<pre class="r"><code>prom &lt;- reduce(promoters, ignore.strand = TRUE)
peaks &lt;- reduce(gr1)
both &lt;- intersect(prom, peaks)
only.prom &lt;- setdiff(prom, both)
only.peaks &lt;- setdiff(peaks, both)
overlapMat &lt;- matrix(0,, ncol = 2, nrow = 2)
colnames(overlapMat) &lt;- c(&quot;in.peaks&quot;, &quot;out.peaks&quot;)
rownames(overlapMat) &lt;- c(&quot;in.promoters&quot;, &quot;out.promoter&quot;)
overlapMat[1,1] &lt;- sum(width(both))
overlapMat[1,2] &lt;- sum(width(only.prom))
overlapMat[2,1] &lt;- sum(width(only.peaks))
overlapMat[2,2] &lt;- 3*10^9 - sum(overlapMat)
round(overlapMat / 10^6, 2)</code></pre>
<pre><code>##              in.peaks out.peaks
## in.promoters     3.02     59.26
## out.promoter     8.17   2929.55</code></pre>
<p>Here we have just used the genome size of 3 billion bases. This is not correct. There are many of these bases which have not been sequenced and in addition, there are many bases which cannot be mapped using short reads. This will reduce the genome size.</p>
<p>Nevertheless, let us compute an odds-ratio for this table:</p>
<pre class="r"><code>oddsRatio &lt;- overlapMat[1,1] * overlapMat[2,2] / (overlapMat[2,1] * overlapMat[1,2])
oddsRatio</code></pre>
<pre><code>## [1] 18.26938</code></pre>
<p>This odds-ratio shows an enrichment of peaks in promoters. We can get a feel for how much the genome size (which we use incorrectly) affects our result by using a lower bound on the genome size. Let us say 1.5 billion bases:</p>
<pre class="r"><code>overlapMat[2,2] &lt;- 1.5*10^9
oddsRatio &lt;- overlapMat[1,1] * overlapMat[2,2] / (overlapMat[2,1] * overlapMat[1,2])
oddsRatio</code></pre>
<pre><code>## [1] 9.354362</code></pre>
<p>The odds-ratio got smaller, but it is still bigger than 1.</p>
<p>Here we basically says that each base can be assigned to peaks or promoters independently, which is definitely false. So there are many, many reasons why this calculation is not the “right” one. Nevertheless, it appears that we have some enrichment, as expected.</p>
</div>
<div id="sessioninfo" class="section level2">
<h2>SessionInfo</h2>
<p></p>
<pre><code>## R version 3.2.1 (2015-06-18)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.10.5 (Yosemite)
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
## [1] AnnotationHub_2.0.3  rtracklayer_1.28.8   GenomicRanges_1.20.5
## [4] GenomeInfoDb_1.4.2   IRanges_2.2.7        S4Vectors_0.6.3     
## [7] BiocGenerics_0.14.0  BiocStyle_1.6.0      rmarkdown_0.7       
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.0                  BiocInstaller_1.18.4        
##  [3] formatR_1.2                  futile.logger_1.4.1         
##  [5] XVector_0.8.0                bitops_1.0-6                
##  [7] futile.options_1.0.0         tools_3.2.1                 
##  [9] zlibbioc_1.14.0              digest_0.6.8                
## [11] evaluate_0.7.2               RSQLite_1.0.0               
## [13] shiny_0.12.2                 DBI_0.3.1                   
## [15] curl_0.9.2                   yaml_2.1.13                 
## [17] stringr_1.0.0                httr_1.0.0                  
## [19] knitr_1.11                   Biostrings_2.36.3           
## [21] Biobase_2.28.0               R6_2.1.0                    
## [23] AnnotationDbi_1.30.1         XML_3.98-1.3                
## [25] BiocParallel_1.2.20          lambda.r_1.1.7              
## [27] magrittr_1.5                 Rsamtools_1.20.4            
## [29] htmltools_0.2.6              GenomicAlignments_1.4.1     
## [31] mime_0.3                     interactiveDisplayBase_1.6.0
## [33] xtable_1.7-4                 httpuv_1.3.3                
## [35] stringi_0.5-5                RCurl_1.95-4.7</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
